File: Educative/SlidingWindow/StringAnagrams.py
# Given a string and a pattern, find all anagrams of the pattern in the given string and return a list of starting indices of the anagrams in the given string.
# Time O(n), Space O(n)

def find_string_anagrams(str1, pattern):
  window_start, matches = 0, 0
  char_freq, indexes = {}, []

  for char in pattern:
    char_freq[char] = char_freq.get(char, 0) + 1

  for window_end in range(len(str1)):
    right_char = str1[window_end]
    if right_char in char_freq:
      char_freq[right_char] -= 1
      if char_freq[right_char] == 0:
        matches += 1

    if matches == len(char_freq):
      indexes.append(window_start)

    if window_end >= len(pattern) - 1:
      left_char = str1[window_start]
      if left_char in char_freq:
        if char_freq[left_char] == 0:
          matches -= 1
        char_freq[left_char] += 1
      window_start += 1

  return indexes

File: Educative/SlidingWindow/StringAnagrams.java
// Given a string and a pattern, find all anagrams of the pattern in the given string and return a list of starting indices of the anagrams in the given string.
// Time O(n), Space O(n)

import java.util.*;

class StringAnagrams {
  public static List<Integer> findStringAnagrams(String str, String pattern) {
    int windowStart = 0, matches = 0;
    Map<Character, Integer> freqMap = new HashMap<>();
    List<Integer> indices = new ArrayList<Integer>();

    for (char chr : pattern.toCharArray())
      freqMap.put(chr, freqMap.getOrDefault(chr, 0) + 1);
    
    for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
      char rightChar = str.charAt(windowEnd);
      if (freqMap.containsKey(rightChar)) {
        freqMap.put(rightChar, freqMap.get(rightChar) - 1);
        if (freqMap.get(rightChar) == 0)
          matches++;
      }

      if (matches == freqMap.size())
        indices.add(windowStart);
      
      if (windowEnd >= pattern.length() - 1) {
        char leftChar = str.charAt(windowStart);
        if (freqMap.containsKey(leftChar)) {
          if (freqMap.get(leftChar) == 0)
            matches--;
          freqMap.put(leftChar, freqMap.get(leftChar) + 1);
        }
        windowStart++;
      }
    }

    return indices;
  }
}